server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4414,4 +4414,33 @@ public void testOpenIndexCreateTranslogKeepOnlyLastCommit() throws Exception {
             assertThat(userData.get(Translog.TRANSLOG_GENERATION_KEY), equalTo("1"));
         }
     }
+
+    public void testCleanUpCommitsWhenGlobalCheckpointAdvanced() throws Exception {
+        IOUtils.close(engine, store);
+        store = createStore();
+        final AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.UNASSIGNED_SEQ_NO);
+        try (InternalEngine engine = createEngine(store, createTempDir(), globalCheckpoint::get)) {
+            final int numDocs = scaledRandomIntBetween(10, 100);
+            for (int docId = 0; docId < numDocs; docId++) {
+                index(engine, docId);
+                if (rarely()) {
+                    globalCheckpoint.set(randomLongBetween(globalCheckpoint.get(), engine.getLocalCheckpointTracker().getCheckpoint()));
```

Comment:
don't we have to use `engine.getLocalCheckpointTracker().getCheckpoint()-1` to make sure the rest goes well and the GCP doesn't go backwards?